repos/iiif-storage/src/Controller/PresleyController.php

```diff
@@ -0,0 +1,168 @@
+                            "",
+                            "",
+                            "",
+                            ""
+                        ]
+                    ];
+                    return $json;
+                }, $manifests),
+            ]
+            , 200);
+    }
+
+    /**
+     * Gets single derived manifest collection by ID
+     */
+    public function getManifestsCollectionAction()
+    {
+        $this->allowCors();
+
+        $collectionId = $this->params()->fromRoute('collection');
+
+        $manifestItem = $this->repo->getById($collectionId);
+        $manifest = $this->builder->build($manifestItem);
+
+        $manifests = $this->repo->getReplacements($collectionId);
+
+        $json = $manifest->getJson();
+        // Invalid on a collection
+        unset($json['structures']);
+        unset($json['sequences']);
+        $json['@type'] = 'sc:Collection';
+        $json['@id'] = $this->url()->fromRoute(null, [], ['force_canonical' => true], true);
+        $json['members'] = array_map(function($manifest) {
+            return $this->builder->build($manifest, false, 0, 1)->getJson();
+        }, $manifests);
+
+        return new JsonResponse($json);
+    }
+
+    /**
+     * Adds derived manifest to the manifests collection
+     */
+    public function addManifestToCollectionAction()
+    {
+        $request = $this->getRequest();
+        if ($request->isOptions()) {
+            return $this->preflightAction();
+        }
+        $this->allowCors();
+
+        $collectionId = $this->params()->fromRoute('collection');
+        $body = json_decode($this->getRequest()->getContent(), true);
+        $label = $body['label'] ?? 'Untitled manifest';
+        $canvasIds = $body['canvasIds'] ?? [];
+
+        $manifestItem = $this->repo->create(function (ItemRequest $item) use ($label, $collectionId, $canvasIds) {
+            $item->addFields(
+                FieldValue::literalsFromRdf('dcterms:title', 'Label', $label)
+            );
+
+            // Set the manifest URI to internal.
+            $item->addField(
+                FieldValue::url('dcterms:identifier', 'Manifest URI', 'uri:internal')
+            );
+
+            $item->addField(
+                FieldValue::entity('dcterms:replaces', $collectionId, 'resource:item')
+            );
+
+            foreach ($canvasIds as $canvasId) {
+                $item->addField(
+                    FieldValue::entity('sc:hasCanvases', $this->extractCanvasId($canvasId))
+                );
+            }
+        });
+
+        $manifest = $this->builder->build($manifestItem);
+
+        return new JsonResponse($manifest->getJson());
+    }
+
+    public function extractCanvasId($id)
+    {
+        if (strpos($id, 'iiif/api') !== false) {
+            $parts = explode('/', array_shift(explode('?', $id)));
+            $id = array_pop($parts);
+            if ($id) {
+                return $id;
+            }
+        }
+        return null;
```

Comment:
This is copied from another part of the code, it's an efficient way to convert Madoc-generated IIIF resources back to their Omeka id. It would be cleaner to use the Zend router to parse it out, but slower and effectively doing the same thing with regex.

Reply:
Can we do this with a regex instead? This `pop(explode($str, shift(explode($str, $val)))` stuff is unreadable.

Reply:
Also, for science: https://gist.github.com/garyttierney/f3267ca99ab10136445c0197ff505c61.

The regex is (unnoticeable) faster.